collections/_tutorials/2022-06-20-tutorial_transformer.md

```diff
@@ -0,0 +1,124 @@
+
+$$z_0 = [x_{class}; \, x^1_p\mathbf{E}; \, x^2_p\mathbf{E}; \, \cdots; \, x^N_p\mathbf{E}] + \mathbf{E}_{pos}$$
+
+with <br>
+&nbsp; &nbsp; &nbsp; &nbsp;$$\mathbf{E}$$ being the linear projection matrix of size $$\mathbf{E} \in \mathbb{R}^{(P^2 \cdot C) \times D}$$ <br>
+&nbsp; &nbsp; &nbsp; &nbsp;$$\mathbf{E}_{pos}$$ being the output of the position embedding operation with dimensions $$\mathbf{E}_{pos} \in \mathbb{R}^{(N+1) \times D}$$
+
+&nbsp;
+
+The figure below illustrates the tokenization process used in ViT.
+
+![](/collections/images/transformers/vit_tokenization.jpg)
+
+&nbsp;
+
+## **Transformer encoder**
+
+The figure below presents an overview of the encoding layer of transformer. It is composed of two main steps.
+
+![](/collections/images/transformers/vit_encoding_layer.jpg)
+
+&nbsp;
+
+The goal of the first step is to compute attention maps between tokens. The output corresponds to $$z^{'}_l$$ which is modeled as follows:
+
+$$z^{'}_l = MSA(LN(z_{l-1})) + z_{l-1} \quad \quad \quad \quad l=1 \cdots L$$
+
+The goal of the second step is to introduce non-linearities to compute more relevant sequence of tokens. The output corresponds to $$z_l$$ which is modeled as follows:
+
+$$z_l = MLP(LN(z^{'}_{l})) + z^{'}_{l} \quad \quad \quad \quad l=1 \cdots L$$
+
+&nbsp;
+
+### Norm step
+
+Normalization was inserted into the encoding layer to control the dynamics of the token values before each of the two key steps. The corresponding diagram is given below:
+
+![](/collections/images/transformers/vit_norm_step.jpg)
+
+&nbsp;
+
+### MLP step
+
+The diagram of the MLP procedure is given below:
+
+![](/collections/images/transformers/vit_mlp_step.jpg)
+
+&nbsp;
+
+### Multi-Head Attention block
+
+The Multi-Head Attention (MHA) block is the key element of the encoding layer. It is based on the "qkv" paradigm, but what does "qkv" mean ? 
+Before answering this question, let's zoom in on the MHA block and get an overview of its structure.
+
+![](/collections/images/transformers/vit_mha_overview.jpg)
+
+From this figure, we can see that the key element is the Self-Attention module which outputs $$k$$ head matrices of size $$\mathbb{R}^{(N+1) \times D_h}$$, where $$D_h$$ is usually computed as $$D_h=D/k$$. These head matrices contain useful information computed from attention mechanisms. The second part of the MHA block uses a linear projection to optimally merge the different heads and to output a new sequence of tokens of the same size as the input matrix, $$i.e.$$ $$\mathbb{R}^{(N+1) \times D}$$. This operation is modeled as:
+
+$$ MSA(z) = \left[ {SA}_1(z), \, {SA}_2(z), \cdots, \, {SA}_k(z) \right] \cdot \mathbf{U}_{msa} \quad \quad \quad \quad \mathbf{U}_{msa} \in \mathbb{R}^{(k D_h) \times D}$$
```

Comment:
Same mix-up between MSA and MHA here? What about the MSA in the subscript of U?

+ formatting improvements.
```suggestion
$$ \text{MHA}(z) = \left[ \text{{SA}}_1(z), \, \text{{SA}}_2(z), \cdots, \, \text{{SA}}_k(z) \right] \cdot \mathbf{U}_{msa} \quad \quad \quad \quad \mathbf{U}_{msa} \in \mathbb{R}^{(k D_h) \times D}$$
```